### PR TITLE
writing updated coord AND xtbopt.xyz during external driver optimization (including nicer printout)

### PR DESCRIPTION
**Pull request by marcelmbn**
- write xtbopt.xyz constantly in the external driver mode so that not only "coord"-based external methods can be used for geometry optimization.
- nicer formatting

### CODE REVIEW

File: src/extern/driver.f90 [hunk 1]
```diff
@@ -122,6 +124,14 @@ subroutine singlepoint(self, env, mol, chk, printlevel, restart, &
     end if
     if (.not. cache) then
       call wrtm(mol%n, mol%at, mol%xyz)
+      write(env%unit,'(/,a)') &
+         "updated geometry written to: coord"
+      call generateFileName(tmpname, 'xtbopt', extension, mol%ftype)
+      write(env%unit,'(a,1x,a,/)') &
+         "updated geometry written to:",tmpname
+      call open_file(ich,tmpname,'w')
+      call writeMolecule(mol, ich, format=1)
```
Comment (awvwgk) [inline src/extern/driver.f90:133]: Why is `format=1` hard-coded here?

File: src/extern/driver.f90 [hunk 2]
```diff
@@ -117,11 +119,22 @@ subroutine singlepoint(self, env, mol, chk, printlevel, restart, &
...
       call rdtm(env,mol%n, .true., energy, gradient, xyz_cached)
       cache = all(abs(xyz_cached - mol%xyz) < 1.e-10_wp)
     end if
     if (.not. cache) then
-      call wrtm(mol%n, mol%at, mol%xyz)
+      call generateFileName(tmpname, 'xtbopt', extension, mol%ftype)
+      write(env%unit,'(/,a,1x,a,/)') &
+         "updated geometry written to:",tmpname
+      call open_file(ich,tmpname,'w')
+      if (exist) then
```
Comment (awvwgk) [inline src/extern/driver.f90:131]: This will only write the gradient if the geometry mismatches the one for the gradient calculation

**Review by awvwgk — CHANGES_REQUESTED** [review]

**Review by MtoLStoN — APPROVED** [review]

File: src/extern/driver.f90 [hunk 3]
```diff
@@ -117,11 +120,27 @@ subroutine singlepoint(self, env, mol, chk, printlevel, restart, &
     !$omp critical (turbo_lock)
     inquire (file='gradient', exist=exist)
     if (exist) then
+      ! ### only TM output is supported for now ###
       call rdtm(env,mol%n, .true., energy, gradient, xyz_cached)
       cache = all(abs(xyz_cached - mol%xyz) < 1.e-10_wp)
+      if (cache) then
+        write(env%unit,'(/,a,/)') &
+         "Geometry is equivalent to the one in &
+         'gradient'. Reading gradient from: 'gradient'."
```
Comment (awvwgk) [inline src/extern/driver.f90:129]: Please guard this with `if (printlevel > 0)` to avoid clutter.

File: src/extern/driver.f90 [hunk 4]
```diff
@@ -117,11 +120,27 @@ subroutine singlepoint(self, env, mol, chk, printlevel, restart, &
...
+      endif
     end if
     if (.not. cache) then
-      call wrtm(mol%n, mol%at, mol%xyz)
+      call generateFileName(tmpname, 'xtbopt', extension, mol%ftype)
+      write(env%unit,'(/,a,1x,a,/)') &
+         "updated geometry written to:",tmpname
```
Comment (awvwgk) [inline src/extern/driver.f90:136]: This output should be guarded by `printlevel > 0`

File: src/extern/driver.f90 [hunk 5]
```diff
@@ -117,11 +120,27 @@ subroutine singlepoint(self, env, mol, chk, printlevel, restart, &
...
       cache = all(abs(xyz_cached - mol%xyz) < 1.e-10_wp)
+      if (cache) then
+        write(env%unit,'(/,a,/)') &
+         "Geometry is equivalent to the one in &
+         'gradient'. Reading gradient from: 'gradient'."
+      endif
     end if
     if (.not. cache) then
-      call wrtm(mol%n, mol%at, mol%xyz)
+      call generateFileName(tmpname, 'xtbopt', extension, mol%ftype)
```
Comment (awvwgk) [inline src/extern/driver.f90:134]: We use `xtbopt` for the final optimized structure, but this routine can be called from a single point or hessian calculation. Better use a more neutral filename here.

File: src/extern/driver.f90 [hunk 6]
```diff
@@ -104,24 +104,48 @@ subroutine singlepoint(self, env, mol, chk, printlevel, restart, &
...
+    cache = .false.
     energy = 0.0_wp
     gradient(:, :) = 0.0_wp
     sigma(:, :) = 0.0_wp
     hlgap = 0.0_wp
     efix = 0.0_wp
     dipole(:) = 0.0_wp
 
+    write(*,*) printlevel
+
```
Comment (MtoLStoN) [inline src/extern/driver.f90:121]: This is probably a debug line, that should be removed.

File: src/extern/driver.f90 [hunk 7]
```diff
@@ -104,24 +104,48 @@ subroutine singlepoint(self, env, mol, chk, printlevel, restart, &
...
     !$omp critical (turbo_lock)
     inquire (file='gradient', exist=exist)
     if (exist) then
+      ! ### only TM output is supported for now ###
       call rdtm(env,mol%n, .true., energy, gradient, xyz_cached)
       cache = all(abs(xyz_cached - mol%xyz) < 1.e-10_wp)
+      if (cache) then
+       if (printlevel > 0) &
```
Comment (MtoLStoN) [inline src/extern/driver.f90:129]: This could be a single if statement instead of a nested one.

**Review by MtoLStoN — CHANGES_REQUESTED** [review]

**Review by awvwgk — APPROVED** [review]
Looks good to me now

**Pull request merged**